pageserver/src/page_service.rs

```diff
@@ -447,6 +539,93 @@ impl PageServerHandler {
         Ok(())
     }
 
+    fn handle_import_basebackup(
+        &self,
+        pgb: &mut PostgresBackend,
+        tenant_id: ZTenantId,
+        timeline_id: ZTimelineId,
+        base_lsn: Lsn,
+        _end_lsn: Lsn,
+    ) -> anyhow::Result<()> {
+        thread_mgr::associate_with(Some(tenant_id), Some(timeline_id));
+        let _enter =
+            info_span!("import basebackup", timeline = %timeline_id, tenant = %tenant_id).entered();
+
+        // Create empty timeline
+        info!("creating new timeline");
+        let repo = tenant_mgr::get_repository_for_tenant(tenant_id)?;
+        let timeline = repo.create_empty_timeline(timeline_id, Lsn(0))?;
+        let repartition_distance = repo.get_checkpoint_distance();
+        let mut datadir_timeline =
+            DatadirTimeline::<LayeredRepository>::new(timeline, repartition_distance);
+
+        // TODO mark timeline as not ready until it reaches end_lsn?
+        // This would prevent compute node from connecting to it and writing conflicting wal.
+
+        // TODO leave clean state on error
+
+        // Import basebackup provided via CopyData
+        info!("importing basebackup");
+        pgb.write_message(&BeMessage::CopyInResponse)?;
+        let reader = CopyInReader::new(pgb);
+        import_basebackup_from_tar(&mut datadir_timeline, reader, base_lsn)?;
+
+        // TODO check checksum
+        // Meanwhile you can verify client-side by taking fullbackup
+        // and checking that it matches in size with what was imported.
+        // It wouldn't work if base came from vanilla postgres though,
+        // since we discard some log files.
+
+        // Flush data to disk, then upload to s3
+        info!("flushing layers");
+        datadir_timeline.tline.checkpoint(CheckpointConfig::Flush)?;
+
+        // TODO Wait for s3 upload to complete
+        // info!("uploading layers");
```

Comment:
This looks important.

If pageserver crashes right after data loading, we will lose the data.
That's less of a problem for bootstrap, because we can re-run initdb anytime.

@LizardWizzard, is there any way to force s3 uploading for the timeline? 
To wait until completion, we must check `remote_consistent_lsn == last_record_lsn`

Reply:
I agree it's important. But afaik it's a general problem, even if we're not importing timelines. Currently GC could eat some layers that are only scheduled for upload, but not uploaded (or something of that sort).

I guess before merging, I'll confirm this is a bigger issue and make a github issue if so.

Reply:
Seems like the condition "The layer has been uploaded" is missing [here](https://github.com/neondatabase/neon/blob/basebackup-import/pageserver/src/layered_repository.rs#L2248), but maybe it's implied from the other conditions somehow?

Reply:
> is there any way to force s3 uploading for the timeline?

We can trigger a checkpoint, using psql for instance: https://github.com/neondatabase/neon/blob/a91e0c299d3604086be9186d31209b08b733766e/test_runner/batch_others/test_remote_storage.py#L73

if it works without errors, it will create a layer that'll schedule its upload to the remote storage and upload it eventually, thus updating the remote Lsn.

We don't have any good way to synchronise the sync part of pageserver with async remote storage part in terms of timeline statuses (neither we have a good timeline status transition impl in pageserver generally).
We register remote storage timeline updates once per remote storage sync loop, placing all changed data into the remote index that can be accessed from anywhere, including the HTTP timeline info endpoint, that can be used to check whether we've synced more data to the remote storage for a certain timeline.

Reply:
I clarified that it's the client's responsibility to wait for upload and implemented that client side in the tests. We can improve on this later, but it's good enough I think.